src/main/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateExecution.java

```diff
@@ -1,33 +1,86 @@
 
     @Inject(optional = true)
-    private transient ExwsAllocateStep exwsAllocateStep;
+    private transient ExwsAllocateStep step;
 
     @StepContextParameter
     private transient Run<?, ?> run;
+    @StepContextParameter
+    private transient TaskListener listener;
 
     @Override
-    protected Void run() throws Exception {
-        List<DiskPool> diskPools = exwsAllocateStep.getDescriptor().getDiskPools();
-        // TODO use globally defined disk pools
-        System.out.println(diskPools);
+    protected ExternalWorkspace run() throws Exception {
+        if (step.getUpstream() == null) {
+            // this is the upstream job
+
+            String diskPoolId = step.getDiskPoolId();
+            if (diskPoolId == null) {
+                throw new Exception("Disk Pool ID was not provided as step parameter");
+            }
+
+            List<DiskPool> diskPools = step.getDescriptor().getDiskPools();
+            DiskAllocationStrategy allocationStrategy = new MostUsableSpaceStrategy(diskPoolId, diskPools);
+            Disk disk = allocationStrategy.allocateDisk();
+
+            String physicalPathOnDisk = disk.getPhysicalPathOnDisk();
+            if (physicalPathOnDisk == null) {
+                String message = format("Physical path on disk was not provided in the Jenkins global config for the Disk Pool ID '%s'", diskPoolId);
+                throw new Exception(message);
+            }
+            String diskId = disk.getDiskId();
+            if (diskId == null) {
+                String message = format("Disk ID was not provided in the Jenkins global config for the Disk Pool ID '%s'", diskPoolId);
+                throw new Exception(message);
+            }
+
+            String pathOnDisk = computePathOnDisk(physicalPathOnDisk);
+            ExternalWorkspace externalWorkspace = new ExternalWorkspace(diskId, pathOnDisk);
+
+            listener.getLogger().println(format("Selected Disk ID is: %s", externalWorkspace.getDiskId()));
+            listener.getLogger().println(format("The path on Disk is: %s", externalWorkspace.getPathOnDisk()));
+
+            return externalWorkspace;
+        } else {
+            // this is the downstream job
+            // TODO implement
+            return null;
+        }
+    }
 
-        // TODO implement method
-        return null;
+    /**
+     * Computes the path to be used on the physical disk.
+     * The computed path is like: physicalPathOnDisk/$JOB_NAME/$BUILD_NUMBER. Where $JOB_NAME also includes all the
+     * folders, if Folders plugin is in use.
+     *
+     * @param physicalPathOnDisk the physical path on the disk
+     * @return the computed file path on the physical disk
+     */
+    private String computePathOnDisk(String physicalPathOnDisk) {
+        FilePath diskFilePath = new FilePath(new File(physicalPathOnDisk));
+        return new FilePath(diskFilePath, run.getParent().getFullName() + "/" + run.getNumber()).getRemote();
```

Comment:
It's even worse. Just keep in mind you may need symbol escaping in the final version